server/py/services/api/runtime_handlers/base.py

```diff
@@ -1751,6 +1751,15 @@ def _ensure_run_state(
             elif run_state == RunStates.error:
                 # Try resolving the error reason
                 reason, message = self._resolve_container_error_status(runtime_resource)
+                # Should run be retried
+                retry_spec = run.get("spec", {}).get("retry")
+                max_retries = retry_spec.get("count") if retry_spec else -1
+                retry_count = run.get("status", {}).get("retry_count", 0) or 0
+                if retry_count < max_retries:
+                    run_state = RunStates.pending_retry
```

Comment:
It might help to log the transition to pendingRetry explicitly here (with run UID and retry count)